### PR TITLE
ENH: 2D compat for DTA tz_localize, to_period

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -80,6 +81,26 @@
 DatetimeLikeArrayT = TypeVar("DatetimeLikeArrayT", bound="DatetimeLikeArrayMixin")
 
 
+def ravel_compat(meth):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:84]: this actually could be in a more general location, maybe pandas/core/array_alogs/ somewhere?

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:84]: maybe core.arraylike?  For now im just excited about getting `__array_function__` working on NDArrayBackedExtensionArray.

**Comment by jbrockmendel** [comment]
btw who should i bug about resample/BinGrouper?

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py:84]: sure

**Comment by jreback** [comment]
> btw who should i bug about resample/BinGrouper?

me or @mroeschke

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:84]: oh, wait, the implementation here is specific to NDArrayBackedExtensionArray, will move it there

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -681,12 +681,13 @@ def value_counts(self, dropna=False):
 
         cls = type(self)
 
-        result = value_counts(values, sort=False, dropna=dropna)
+        result = value_counts(values.ravel("K"), sort=False, dropna=dropna)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimelike.py:684]: What's the reason for the ravel here? (value_counts is always done on a 1D array AFAIK)

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
Can you explain the rationale?

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: moved + green

**Comment by jbrockmendel** [comment]
> Can you explain the rationale?

Two short-term motivations.  First, a number of places in DatetimeLikeBlock would be simpler if we just stored the EA as self.values, which requires more robust 2D support.  Second, I've got a branch implementing `__array_function__` and got some failures (dont remember which off the top of my head) that pushed this up the priority list.

**Comment by jorisvandenbossche** [comment]
> a number of places in DatetimeLikeBlock would be simpler if we just stored the EA as self.values, which requires more robust 2D support

Can you give a bit more details about which places that would be? 
I am personally -1 on storing the EA on 2D blocks. The 2D EA was a compromise for the block-wise ops, but expanding its usage first requires discussion IMO.

> Second, I've got a branch implementing `__array_function__` and got some failures

How is `to_period` or `tz_localize` related to `__array_function__`?

Also, IMO we should only add this for those functions for which we actually start using it. None of the changes here are actually already used somewhere?

**Comment by jbrockmendel** [comment]
>> a number of places in DatetimeLikeBlock would be simpler if we just stored the EA as self.values, which requires more robust 2D support

> Can you give a bit more details about which places that would be?

Most of these have `# TODO(EA2D):` comments: iget, diff, shift, set_inplace, get_values, quantile, fillna

**Comment by jorisvandenbossche** [comment]
> Most of these have # TODO(EA2D): comments: iget, diff, shift, set_inplace, get_values, quantile, fillna

I know you have been adding a lot of those comments, but having those comments doesn't mean there is consensus on further expanding the 2D capability of the datetimelike arrays ..

We could also make the DatetimeBlock a 1D non-consolidatable block, and that would also solve the datetime-specific TODO(EA2D) comments

**Comment by jbrockmendel** [comment]
> We could also make the DatetimeBlock a 1D non-consolidatable block

We're going to do essentially that in ArrayManager, which we're implementing so we can try it out and see what the performance/complexity impact is.  Fleshing out the 2D support for datetimelike arrays is part of making it possible to also try out the EA2D approach.

**Comment by jreback** [comment]
> > We could also make the DatetimeBlock a 1D non-consolidatable block
> 
> We're going to do essentially that in ArrayManager, which we're implementing so we can try it out and see what the performance/complexity impact is. Fleshing out the 2D support for datetimelike arrays is part of making it possible to also try out the EA2D approach.

I agree here. the array manager perf implications will determine our ultimate direction. e.g. if we get good performance then we should certainly move everything as it makes things code-wise way simpler. if not we have to re-think and potentially support 2d EA.

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py:691]: I agree this one cannot be ravel_compat (instead assert values.ndim == 1 here)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: updated+green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**